src/main/java/org/opensearch/plugin/insights/core/service/grouper/QueryGrouper.java

```diff
@@ -0,0 +1,66 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.plugin.insights.core.service.grouper;
+
+import org.opensearch.plugin.insights.core.service.store.PriorityQueueTopQueriesStore;
+import org.opensearch.plugin.insights.rules.model.GroupingType;
+import org.opensearch.plugin.insights.rules.model.SearchQueryRecord;
+
+/**
+ * Interface for grouping search queries based on grouping type for the metric type.
+ */
+public interface QueryGrouper {
+
+    /**
+     * Add query to the group based on the GroupType setting.
+     * @param searchQueryRecord record to be added
+     * @return the aggregate search query record representing the group
+     */
+    SearchQueryRecord addQueryToGroup(SearchQueryRecord searchQueryRecord);
+
+    /**
+     * Drain the internal grouping. Needs to be performed after every window or if a setting is changed.
+     */
+    void drain();
+
+    /**
+     * Get the min heap queue that holds the top N queries.
+     *
+     * @return the PriorityBlockingQueue containing the top N queries
+     */
+    PriorityQueueTopQueriesStore<SearchQueryRecord> getMinHeapTopQueriesStore();
```

Comment:
Same as above. Need to code to interfaces!
PriorityQueueTopQueriesStore -> TopQueriesStore

getMinHeapTopQueriesStore ->  getTopQueriesStore

Need to change javadoc as well accordingly.